File: backend/app/episodes.py
from __future__ import annotations

import bisect
from datetime import timedelta
from typing import List
import re

from sqlalchemy import and_, select
from sqlalchemy.orm import aliased, Session

from . import models, schemas
from .llm import summarize_episode, summarize_file_evolution


def _parse_issue_numbers(text: str) -> list[int]:
    return [int(m) for m in re.findall(r"#(\d+)", text or "")]


def _build_llm_context(
    window: list[models.Commit],
    pr: models.PullRequest | None,
    issue: models.Issue | None,
) -> dict:
    return {
        "pr_title": pr.title if pr else "",
        "pr_body": (pr.body[:600] if pr and pr.body else ""),
        "commit_messages": "\n".join(c.message or "" for c in window)[:600],
        "issue_title": issue.title if issue else "",
        "issue_body": (issue.body[:400] if issue and issue.body else ""),
    }


def build_and_persist_episodes(repo_id: int, db: Session) -> None:
    """
    Groups commits per file into episodes (by PR first, then time-windowed for
    raw commits), calls LLM per episode, and persists Episode + EpisodeMember
    rows. Safe to re-run (clears old rows first).
    """
    # Clear existing episodes for this repo
    old_eps = db.query(models.Episode).filter_by(repo_id=repo_id).all()
    for ep in old_eps:
        db.query(models.EpisodeMember).filter_by(episode_id=ep.id).delete()
        db.delete(ep)
    db.commit()

    # Fetch all commits for this repo ordered by date
    commits = (
        db.query(models.Commit)
        .filter_by(repo_id=repo_id)
        .filter(models.Commit.date.isnot(None))
        .order_by(models.Commit.date.asc())
        .all()
    )
    if not commits:
        return

    # ── Pre-load merged PRs once (avoids N+1 in the heuristic loop below) ──
    merged_prs = (
        db.query(models.PullRequest)
        .filter_by(repo_id=repo_id)
        .filter(models.PullRequest.merged_at.isnot(None))
        .order_by(models.PullRequest.merged_at.asc())
        .all()
    )
    merged_pr_times = [pr.merged_at for pr in merged_prs]

    # ── Pre-load PR id → object map for direct PR lookups ──────────────────
    pr_by_id: dict[int, models.PullRequest] = {pr.id: pr for pr in merged_prs}
    # Also include non-merged PRs (linked commits may reference open PRs)
    all_prs = (
        db.query(models.PullRequest)
        .filter_by(repo_id=repo_id)
        .all()
    )
    for pr in all_prs:
        pr_by_id.setdefault(pr.id, pr)

    # ── Episode grouping ────────────────────────────────────────────────────
    windows: list[list[models.Commit]] = []

    commits_by_pr: dict[int, list[models.Commit]] = {}
    unlinked_commits: list[models.Commit] = []

    for c in commits:
        if c.pr_id is not None:
            commits_by_pr.setdefault(c.pr_id, []).append(c)
        else:
            unlinked_commits.append(c)

    for pr_id, pr_commits in commits_by_pr.items():
        windows.append(pr_commits)

    if unlinked_commits:
        window = [unlinked_commits[0]]
        window_end = unlinked_commits[0].date

        for c in unlinked_commits[1:]:
            if c.date - window_end <= timedelta(days=1):
                window.append(c)
                window_end = c.date
            else:
                windows.append(window)
                window = [c]
                window_end = c.date
        windows.append(window)

    windows.sort(key=lambda w: w[0].date)

    for window_commits in windows:
        pr: models.PullRequest | None = None
        issue: models.Issue | None = None

        # (a) Direct pr_id link
        linked_commit = next((c for c in window_commits if c.pr_id is not None), None)
        if linked_commit is not None:
            pr = pr_by_id.get(linked_commit.pr_id)

        # (b) Heuristic: use preloaded list + bisect — O(log N) per window,
        #     not one DB query per commit (was N+1).
        if pr is None:
            approx_hours_after_commit = 24
            for c in window_commits:
                lo = bisect.bisect_left(merged_pr_times, c.date)
                hi = bisect.bisect_right(
                    merged_pr_times, c.date + timedelta(hours=approx_hours_after_commit)
                )
                candidates = merged_prs[lo:hi]
                if not candidates:
                    continue

                pr_match = candidates[0]
                c.pr_id = pr_match.id
                db.flush()
                pr = pr_match
                pr_by_id[pr_match.id] = pr_match
                break

        # Find linked issue via #number mentions
        all_text = " ".join([
            *(c.message or "" for c in window_commits),
            pr.title if pr else "",
            pr.body if pr and pr.body else "",
        ])
        issue_numbers = _parse_issue_numbers(all_text)
        if issue_numbers:
            issue = (
                db.query(models.Issue)
                .filter_by(repo_id=repo_id, number=issue_numbers[0])
                .first()
            )

        if pr:
            title = f"PR #{pr.number}: {pr.title or 'No Title'}"
        elif issue:
            title = f"Issue #{issue.number}: {issue.title or 'No Title'}"
        else:
            title = f"Changes on {window_commits[0].date.date().isoformat()}"

        context = _build_llm_context(window_commits, pr, issue)
        llm_summary = summarize_episode(context)

        ep = models.Episode(
            repo_id=repo_id,
            title=title,
            start_date=window_commits[0].date,
            end_date=window_commits[-1].date,
            llm_summary=llm_summary,
        )
        db.add(ep)
        db.flush()

        for c in window_commits:
            db.add(models.EpisodeMember(episode_id=ep.id, commit_id=c.id, member_type="commit"))
        if pr:
            db.add(models.EpisodeMember(episode_id=ep.id, pr_id=pr.id, member_type="pr"))
        if issue:
            db.add(models.EpisodeMember(episode_id=ep.id, issue_id=issue.id, member_type="issue"))

    db.commit()


def episodes_for_file(repo_id: int, file_path: str, db: Session) -> List[schemas.EpisodeSummary]:
    """
    Reads persisted episodes that touch the given file, returns them sorted by date.
    """
    commit_ids_for_file = (
        select(models.Commit.id)
        .join(models.FileChange, models.FileChange.commit_id == models.Commit.id)
        .where(
            models.Commit.repo_id == repo_id,
            models.FileChange.file_path == file_path,
        )
        .subquery()
    )

    episode_ids_for_file = (
        select(models.EpisodeMember.episode_id)
        .where(
            models.EpisodeMember.commit_id.in_(commit_ids_for_file),
            models.EpisodeMember.member_type == "commit",
        )
        .distinct()
    )

    pr_member = aliased(models.EpisodeMember)
    issue_member = aliased(models.EpisodeMember)

    stmt = (
        select(
            models.Episode.id,
            models.Episode.title,
            models.Episode.start_date,
            models.Episode.end_date,
            models.Episode.llm_summary,
            models.PullRequest.number.label("pr_number"),
            models.Issue.number.label("issue_number"),
        )
        .outerjoin(
            pr_member,
            and_(
                pr_member.episode_id == models.Episode.id,
                pr_member.member_type == "pr",
            ),
        )
        .outerjoin(models.PullRequest, models.PullRequest.id == pr_member.pr_id)
        .outerjoin(
            issue_member,
            and_(
                issue_member.episode_id == models.Episode.id,
                issue_member.member_type == "issue",
            ),
        )
        .outerjoin(models.Issue, models.Issue.id == issue_member.issue_id)
        .where(models.Episode.id.in_(episode_ids_for_file))
        .order_by(models.Episode.start_date.asc())
    )

    rows = db.execute(stmt).all()
    return [
        schemas.EpisodeSummary(
            id=row.id,
            title=row.title,
            start_date=row.start_date,
            end_date=row.end_date,
            llm_summary=row.llm_summary,
            pr_number=row.pr_number,
            issue_number=row.issue_number,
        )
        for row in rows
    ]


def file_story_for_file(repo_id: int, file_path: str, db: Session) -> str | None:
    """
    Build 1-2 sentence high-level story for a file based on chronological episode summaries.
    Returns None when there are no episodes or when the LLM call fails.
    """
    ep_summaries = episodes_for_file(repo_id=repo_id, file_path=file_path, db=db)
    chronological = [ep.llm_summary for ep in ep_summaries if ep.llm_summary]
    if not chronological:
        return None

    story = summarize_file_evolution(episodes_summaries=chronological)
    if not story:
        return None
    if story.startswith("File story failed:"):
        return None
    return story

File: src/episodes.ts
import { Database } from "sql.js";
import { dbAll, dbGet, dbRun, dbRunGetId, persist } from "./db";
import { summarizeEpisode, EpisodeContext } from "./llm";
import { makeLimiter } from "./github";

interface CommitRow {
  id: number; sha: string; date: string; message: string | null; pr_id: number | null;
}
interface PRRow {
  id: number; number: number; title: string | null; body: string | null; merged_at: string | null;
}
interface IssueRow {
  id: number; number: number; title: string | null; body: string | null;
}

function parseIssueNumbers(text: string): number[] {
  return [...text.matchAll(/#(\d+)/g)].map(m => parseInt(m[1]));
}

export interface EpisodeSummary {
  id: number; title: string;
  start_date: string; end_date: string;
  llm_summary: string | null;
  pr_number: number | null; issue_number: number | null;
}

export async function buildAndPersistEpisodes(
  repoId: number, db: Database, apiKey: string, model: string
): Promise<void> {
  // Clear existing episodes
  const oldEps = dbAll(db, "SELECT id FROM episodes WHERE repo_id = ?", repoId);
  for (const ep of oldEps) {
    dbRun(db, "DELETE FROM episode_members WHERE episode_id = ?", ep["id"]);
    dbRun(db, "DELETE FROM episodes WHERE id = ?", ep["id"]);
  }

  const commits = dbAll(
    db,
    "SELECT id, sha, date, message, pr_id FROM commits WHERE repo_id = ? AND date IS NOT NULL ORDER BY date ASC",
    repoId
  ) as unknown as CommitRow[];
  if (!commits.length) { return; }

  const allPRs = dbAll(
    db, "SELECT id, number, title, body, merged_at FROM pull_requests WHERE repo_id = ?", repoId
  ) as unknown as PRRow[];
  const prById = new Map(allPRs.map(pr => [pr.id, pr]));
  const mergedPRs = allPRs
    .filter(pr => pr.merged_at)
    .sort((a, b) => a.merged_at!.localeCompare(b.merged_at!));

  // Group commits by PR, collect unlinked
  const commitsByPr = new Map<number, CommitRow[]>();
  const unlinked: CommitRow[] = [];
  for (const c of commits) {
    if (c.pr_id !== null) {
      const arr = commitsByPr.get(c.pr_id) ?? [];
      arr.push(c);
      commitsByPr.set(c.pr_id, arr);
    } else {
      unlinked.push(c);
    }
  }

  // Heuristic: link unlinked commits to merged PR within 24h
  for (const c of unlinked) {
    const cTime = new Date(c.date).getTime();
    const match = mergedPRs.find(pr => {
      const mergedTime = new Date(pr.merged_at!).getTime();
      return mergedTime >= cTime && mergedTime - cTime <= 86_400_000;
    });
    if (match) {
      c.pr_id = match.id;
      dbRun(db, "UPDATE commits SET pr_id = ? WHERE id = ?", match.id, c.id);
      const arr = commitsByPr.get(match.id) ?? [];
      arr.push(c);
      commitsByPr.set(match.id, arr);
    }
  }
  const stillUnlinked = unlinked.filter(c => c.pr_id === null);

  // Build windows: PR-grouped + time-windowed unlinked
  const windows: { commits: CommitRow[]; prId: number | null }[] = [
    ...[...commitsByPr.entries()].map(([prId, cs]) => ({ commits: cs, prId })),
  ];
  if (stillUnlinked.length) {
    let w = [stillUnlinked[0]];
    for (let i = 1; i < stillUnlinked.length; i++) {
      const gap = new Date(stillUnlinked[i].date).getTime() - new Date(w[w.length - 1].date).getTime();
      if (gap <= 86_400_000) { w.push(stillUnlinked[i]); }
      else { windows.push({ commits: w, prId: null }); w = [stillUnlinked[i]]; }
    }
    windows.push({ commits: w, prId: null });
  }
  windows.sort((a, b) => a.commits[0].date.localeCompare(b.commits[0].date));

  // Build episode summaries in parallel (capped at 3)
  const epLimiter = makeLimiter(3);
  const tasks = windows.map(w => epLimiter(async () => {
    const pr = w.prId !== null ? (prById.get(w.prId) ?? null) : null;
    const allText = [
      ...w.commits.map(c => c.message ?? ""),
      pr?.title ?? "",
      pr?.body ?? "",
    ].join(" ");

    const issueNums = parseIssueNumbers(allText);
    const issue = issueNums.length
      ? (dbGet(db, "SELECT id, number, title, body FROM issues WHERE repo_id = ? AND number = ?",
          repoId, issueNums[0]) as unknown as IssueRow | undefined) ?? null
      : null;

    let title: string;
    if (pr) { title = `PR #${pr.number}: ${pr.title ?? "No Title"}`; }
    else if (issue) { title = `Issue #${issue.number}: ${issue.title ?? "No Title"}`; }
    else { title = `Changes on ${w.commits[0].date.slice(0, 10)}`; }

    const ctx: EpisodeContext = {
      pr_title: pr?.title ?? "",
      pr_body: (pr?.body ?? "").slice(0, 600),
      commit_messages: w.commits.map(c => c.message ?? "").join("\n").slice(0, 600),
      issue_title: issue?.title ?? "",
      issue_body: (issue?.body ?? "").slice(0, 400),
    };

    const llmSummary = await summarizeEpisode(ctx, apiKey, model);
    const sorted = [...w.commits].sort((a, b) => a.date.localeCompare(b.date));

    const epId = dbRunGetId(
      db,
      "INSERT INTO episodes (repo_id, title, start_date, end_date, llm_summary) VALUES (?, ?, ?, ?, ?)",
      repoId, title, sorted[0].date, sorted[sorted.length - 1].date, llmSummary
    );

    for (const c of w.commits) {
      dbRun(db, "INSERT INTO episode_members (episode_id, commit_id, pr_id, issue_id, member_type) VALUES (?, ?, ?, ?, ?)",
        epId, c.id, null, null, "commit");
    }
    if (pr) {
      dbRun(db, "INSERT INTO episode_members (episode_id, commit_id, pr_id, issue_id, member_type) VALUES (?, ?, ?, ?, ?)",
        epId, null, pr.id, null, "pr");
    }
    if (issue) {
      dbRun(db, "INSERT INTO episode_members (episode_id, commit_id, pr_id, issue_id, member_type) VALUES (?, ?, ?, ?, ?)",
        epId, null, null, issue.id, "issue");
    }
  }));

  await Promise.all(tasks);
  persist();
}

export function episodeForCommit(commitId: number, db: Database): EpisodeSummary | null {
  const row = dbGet(db, `
    SELECT e.id, e.title, e.start_date, e.end_date, e.llm_summary,
           pr.number AS pr_number, i.number AS issue_number
    FROM episodes e
    JOIN episode_members em ON em.episode_id = e.id AND em.commit_id = ? AND em.member_type = 'commit'
    LEFT JOIN episode_members epr ON epr.episode_id = e.id AND epr.member_type = 'pr'
    LEFT JOIN pull_requests pr ON pr.id = epr.pr_id
    LEFT JOIN episode_members ei ON ei.episode_id = e.id AND ei.member_type = 'issue'
    LEFT JOIN issues i ON i.id = ei.issue_id
    LIMIT 1
  `, commitId);

  if (!row) { return null; }
  return {
    id: row["id"] as number,
    title: row["title"] as string,
    start_date: row["start_date"] as string,
    end_date: row["end_date"] as string,
    llm_summary: row["llm_summary"] as string | null,
    pr_number: row["pr_number"] as number | null,
    issue_number: row["issue_number"] as number | null,
  };
}